packages/flutter/test/material/text_field_test.dart

```diff
@@ -11541,4 +11541,178 @@ void main() {
+            defaultTextField.loupeBuilder!(
+                context,
+                LoupeController(),
+                ValueNotifier<LoupeSelectionOverlayInfoBearer>(
+                  const LoupeSelectionOverlayInfoBearer.empty(),
+                )),
+            isNull);
+      },
+          variant: TargetPlatformVariant.all(excluding: <TargetPlatform>{
+            TargetPlatform.iOS,
+            TargetPlatform.android
+          }));
+    });
+  });
+
+  group('loupe', () {
+    late ValueNotifier<LoupeSelectionOverlayInfoBearer> infoBearer;
+    final Widget fakeLoupe = Container(key: UniqueKey());
+
+    testWidgets(
+        'Can drag handles to show, unshow, and update loupe',
+        (WidgetTester tester) async {
+      final TextEditingController controller = TextEditingController();
+      await tester.pumpWidget(
+        overlay(
+          child: TextField(
+            dragStartBehavior: DragStartBehavior.down,
+            controller: controller,
+            loupeBuilder: (_,
+                LoupeController controller,
+                ValueNotifier<LoupeSelectionOverlayInfoBearer>
+                    localInfoBearer) {
+              infoBearer = localInfoBearer;
+              return fakeLoupe;
+            },
+          ),
+        ),
+      );
+
+      const String testValue = 'abc def ghi';
+      await tester.enterText(find.byType(TextField), testValue);
+      await skipPastScrollingAnimation(tester);
+
+      // Double tap the 'e' to select 'def'.
+      await tester.tapAt(textOffsetToPosition(tester, testValue.indexOf('e')));
+      await tester.pump(const Duration(milliseconds: 30));
+      await tester.tapAt(textOffsetToPosition(tester, testValue.indexOf('e')));
+      await tester.pump(const Duration(milliseconds: 30));
+
+      final TextSelection selection = controller.selection;
+
+      final RenderEditable renderEditable = findRenderEditable(tester);
+      final List<TextSelectionPoint> endpoints = globalize(
+        renderEditable.getEndpointsForSelection(selection),
+        renderEditable,
+      );
+
+      // Drag the right handle 2 letters to the right.
+      final Offset handlePos = endpoints.last.point + const Offset(1.0, 1.0);
+      final TestGesture gesture = await tester.startGesture(handlePos, pointer: 7);
+
+      Offset? firstDragGesturePosition;
+
+      await gesture.moveTo(textOffsetToPosition(tester, testValue.length - 2));
+      await tester.pump();
+
+      expect(find.byKey(fakeLoupe.key!), findsOneWidget);
+      firstDragGesturePosition = infoBearer.value.globalGesturePosition;
+
+      await gesture.moveTo(textOffsetToPosition(tester, testValue.length));
+      await tester.pump();
+
+      // Expect the position the loupe gets to have moved.
+      expect(firstDragGesturePosition,
+          isNot(infoBearer.value.globalGesturePosition));
+
+      await gesture.up();
+      await tester.pump();
+
+      expect(find.byKey(fakeLoupe.key!), findsNothing);
+    });
```

Comment:
Also, is there anything else that you want to test fully "end to end" like this?  Like gestures that go out of bounds, etc.  I know stuff like that is already covered by more narrow tests elsewhere.  Up to you depending on if you think it would be beneficial.